File: day-2/python/interpreter.py
"""Class representing the interpreter."""

from enum import Enum
from typing import List


class Opcode(Enum):
    """Class for representing intcode methods, mapping opname to opcode."""

    ADD = 1
    MULTIPLY = 2
    HALT = 99


class Interpreter:
    """Class representing the interpreter for IntCode."""

    def __init__(self: "Interpreter", codes: List[int]) -> None:
        """Instantiate a new interpreter given code."""
        self.position = 0
        self.codes = codes
        self.halted = False

    def reset(self: "Interpreter") -> None:
        """Reset the interpreter."""
        self.position = 0
        self.halted = False

    def step(self: "Interpreter") -> None:
        """Step to next instruction, fetching opcode, operands and result."""
        code = Opcode(self.codes[self.position])

        if code == Opcode.HALT:
            self.halted = True
            return

        operand_1 = self.codes[self.codes[self.position + 1]]
        operand_2 = self.codes[self.codes[self.position + 2]]

        if code == Opcode.MULTIPLY:
            result = operand_1 * operand_2
        else:
            result = operand_1 + operand_2

        self.codes[self.codes[self.position + 3]] = result

        self.position += 4

File: day-2/rust/src/interpreter.rs
#[derive(Debug)]
pub struct Interpreter {
    pub codes: Vec<i64>,
    position: i64,
    pub is_running: bool
}

#[derive(Debug, PartialEq)]
pub enum OpCode {
    Add,
    Multiply,
    Halt
}

impl Interpreter {
    pub fn new(codes: std::vec::Vec<i64>) -> Interpreter {
        Interpreter { codes, position: 0, is_running: true }
    }

    pub fn step(&mut self) {
        let op = self.parse_opcode(self.fetch(self.position));

        if op == OpCode::Halt {
            self.is_running = false;
            return;
        }

        let operand_1 = self.get_operand(self.position + 1);
        let operand_2 = self.get_operand(self.position + 2);

        let result = match op {
            OpCode::Add => {
                operand_1 + operand_2
            },
            OpCode::Multiply => {
                operand_1 * operand_2
            },
            OpCode::Halt => {
                0
            }
        };

        self.put(self.fetch(self.position + 3), result);

        self.position += 4;
    }

    fn get_operand(&self, pos: i64) -> i64 {
        self.fetch(self.fetch(pos))
    }

    fn put(&mut self, pos: i64, data: i64) {
        self.codes[pos as usize] = data;
    }

    pub fn fetch(&self, pos: i64) -> i64 {
        self.codes[pos as usize]
    }

    fn parse_opcode(&self, op: i64) -> OpCode {
        match op {
            1 => OpCode::Add,
            2 => OpCode::Multiply,
            99 => OpCode::Halt,
            _ => panic!("Unimplemented opcode: {}", op)
        }
    }
}
